flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesStateHandleStore.java

```diff
@@ -0,0 +1,422 @@
+
+	private static final Logger LOG = LoggerFactory.getLogger(KubernetesStateHandleStore.class);
+
+	private final FlinkKubeClient kubeClient;
+
+	private final String configMapName;
+
+	private final RetrievableStateStorageHelper<T> storage;
+
+	private final Predicate<String> filter;
+
+	private final String lockIdentity;
+
+	private final Supplier<Exception> configMapNotExistSupplier;
+
+	/**
+	 * Creates a {@link KubernetesStateHandleStore}.
+	 *
+	 * @param kubeClient    The Kubernetes client.
+	 * @param storage       To persist the actual state and whose returned state handle is then written to ConfigMap
+	 * @param configMapName ConfigMap to store the state handle store pointer
+	 * @param filter filter to get the expected keys for state handle
+	 * @param lockIdentity lock identity of current HA service
+	 */
+	public KubernetesStateHandleStore(
+			FlinkKubeClient kubeClient,
+			String configMapName,
+			RetrievableStateStorageHelper<T> storage,
+			Predicate<String> filter,
+			String lockIdentity) {
+
+		this.kubeClient = checkNotNull(kubeClient, "Kubernetes client");
+		this.storage = checkNotNull(storage, "State storage");
+		this.configMapName = checkNotNull(configMapName, "ConfigMap name");
+		this.filter = checkNotNull(filter);
+		this.lockIdentity = checkNotNull(lockIdentity, "Lock identity of current HA service");
+
+		this.configMapNotExistSupplier = () -> new KubernetesException("ConfigMap " + configMapName + " not exists.");
+	}
+
+	/**
+	 * Creates a state handle, stores it in ConfigMap. We could guarantee that only the leader could update the
+	 * ConfigMap. Since “Get(check the leader)-and-Update(write back to the ConfigMap)” is a
+	 * transactional operation.
+	 *
+	 * @param key Key in ConfigMap
+	 * @param state State to be added
+	 *
+	 * @throws AlreadyExistException if the name already exists
+	 * @throws Exception if persisting state or writing state handle failed
+	 */
+	@Override
+	public RetrievableStateHandle<T> add(String key, T state) throws Exception {
+		checkNotNull(key, "Key in ConfigMap.");
+		checkNotNull(state, "State.");
+
+		final RetrievableStateHandle<T> storeHandle = storage.store(state);
+
+		boolean success = false;
+
+		try {
+			final byte[] serializedStoreHandle = InstantiationUtil.serializeObject(storeHandle);
+			success = kubeClient.checkAndUpdateConfigMap(
+				configMapName,
+				c -> {
+					if (KubernetesLeaderElector.hasLeadership(c, lockIdentity)) {
+						if (!c.getData().containsKey(key)) {
+							c.getData().put(key, Base64.getEncoder().encodeToString(serializedStoreHandle));
+							return Optional.of(c);
+						} else {
+							throw new CompletionException(new StateHandleStore.AlreadyExistException(
+								key + " already exists in ConfigMap " + configMapName));
+						}
+					}
+					return Optional.empty();
+				}).get();
+			return storeHandle;
+		} catch (Exception ex) {
+			ExceptionUtils.rethrowException(ExceptionUtils.stripCompletionException(ex));
```

Comment:
Since the exception could be wrapped(e.g. `RetryException`), I will use the following code to throw the correct exception.

```
throw ExceptionUtils.findThrowable(ex, NotExistException.class).orElseThrow(() -> ex);
```